plugins/basic/nexus-repository-docker/src/main/java/org/sonatype/nexus/repository/docker/internal/DockerFacetImpl.java

```diff
@@ -0,0 +1,669 @@
+      }
+    }
+    // TODO check that parent asset is accessible via authorisation once we have per assets auth
+    if (parentAsset != null) {
+      ancestry = parentAsset.formatAttributes().get(P_ANCESTRY, TypeTokens.LIST_STRING);
+    }
+    if (ancestry == null) {
+      ancestry = Lists.newArrayList();
+    }
+    ancestry.add(0, layerId);
+
+    Asset metadataAsset = findAsset(tx, bucket, metadataAssetName);
+    if (metadataAsset == null) {
+      metadataAsset = tx.createAsset(bucket, getRepository().getFormat());
+      metadataAsset.name(metadataAssetName);
+      metadataAsset.formatAttributes().set(P_PATH, "/v1/images/" + metadataAssetName);
+      metadataAsset.formatAttributes().set(P_IN_PROGRESS, true);
+    }
+    else {
+      if (!Boolean.TRUE.equals(metadataAsset.formatAttributes().get(P_IN_PROGRESS))) {
+        throw new V1Exception("Image already exists", HttpStatus.CONFLICT);
+      }
+      // cleanup checksums as we have potentially new metadata/layer
+      metadataAsset.formatAttributes().remove(P_X_DOCKER_CHECKSUM_PAYLOAD);
+    }
+    metadataAsset.formatAttributes().set(P_ANCESTRY, ancestry);
+    metadataAsset.formatAttributes().set(P_ARCH, data.get("os"));
+    metadataAsset.formatAttributes().set(P_OS, data.get("architecture"));
+
+    // If we reach that point, it means that this is a new image or a retry on a failed push
+    AttributesMap contentAttributes = null;
+    if (payload instanceof Content) {
+      contentAttributes = ((Content) payload).getAttributes();
+    }
+    Content.applyToAsset(metadataAsset, Content.maintainLastModified(metadataAsset, contentAttributes));
+    tx.setBlob(metadataAsset, metadataAssetName, streamSupplier.get(), hashAlgorithms, null, payload.getContentType());
+    tx.saveAsset(metadataAsset);
+  }
+
+  @Override
+  @Transactional(retryOn = IllegalStateException.class)
+  public Payload getLayerContent(final String layerId) throws IOException {
+    validateLayerId(layerId);
+    StorageTx tx = UnitOfWork.currentTransaction();
+
+    String metadataAssetName = layerId + "/json";
+    String layerAssetName = layerId + "/layer";
+
+    Bucket bucket = tx.getBucket();
+    // check that we have the layer and layer content was fully uploaded
+    Asset metadataAsset = findAsset(tx, bucket, metadataAssetName);
+    if (metadataAsset == null || Boolean.TRUE.equals(metadataAsset.formatAttributes().get(P_IN_PROGRESS))) {
+      throw new V1Exception("Image not found", HttpStatus.NOT_FOUND);
+    }
+    // check that we have layer content
+    Asset layerAsset = findAsset(tx, bucket, layerAssetName);
+    if (layerAsset == null) {
+      throw new V1Exception("Image not found", HttpStatus.NOT_FOUND);
+    }
+
+    return marshall(layerAsset, tx.requireBlob(layerAsset.requireBlobRef()));
+  }
+
+  @Override
+  public void putLayerContent(final String layerId, final Payload content) throws IOException, InvalidContentException {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove the declaration of thrown exception 'org.sonatype.nexus.repository.InvalidContentException' which is a runtime exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)